### ISSUE TITLE
Add Portfolio Solver

### ISSUE THREAD

**Issue by PhilippWendler** [sosy-lab/java-smt#8]
A portfolio solver that delegates to an arbitrary subset of other solvers could be implemented.
For all cheap methods like formula construction, the delegation would be done in a sequential loop. For expensive methods like solving, this would be done in parallel for each solver, waiting until the first one returns an answer and then asking all the others to stop.

After stopping, we would probably need to wait for all solvers to really finish to leave them in a clean state for the next query.

`allSat` is a problem due to the callbacks, but could simply be left unsupported or delegated to one solver only.

Formula traversal done by client code is not possible, but we can add methods for the most common purposes (like extract atoms), and a formula-transformation visitor would still be possible (if it is stateless).

**Comment by stieglma** [comment]
As well as for ```allsat``` this needs to be done for all kinds of ```getInterpolants``` calls, as they can then be done only on the solver that returned UNSAT first.

Why can formula traversals not be done? Because we do not know with which solver they were created? A solution would be to only pass formulas to the user that are created with a certain solver, e.g. the first one of those that should be used.

**Comment by PhilippWendler** [comment]
> As well as for allsat this needs to be done for all kinds of getInterpolants calls, as they can then be done only on the solver that returned UNSAT first.

Yes.

> Why can formula traversals not be done? Because we do not know with which solver they were created?

Because the formula structure may be different for each solver.

> A solution would be to only pass formulas to the user that are created with a certain solver, e.g. the first one of those that should be used.

No, that would defeat the whole purpose of the portfolio solver. If the user has a formula at hand, he ought to be able to use this formula in a query, and all solvers should be used for that query. Thus each formula given to the user must be present in all solvers.

This btw. means that the result of `getInterpolant` needs to be replicated in each of the other solvers via string dumping and parsing. Depending on the solver this may require that we have waited for termination of the previous `isUnsat` call, otherwise we would use methods of one solver in parallel.

**Comment by cheshire** [comment]
Doesn't look like this is coming anytime soon :P

**Comment by PhilippWendler** [comment]
Why close this issue? Just because nobody is working on it, or is the feature rejected?

**Comment by cheshire** [comment]
Yes, as there would be no progress in the foreseeable future. Personally I'd prefer not to keep zombie eventually-nice-to-have tickets, and I'll much rather have the issue re-opened by someone wishing to work on that (or a new one created). I think the discussion above is too short to justify having a zombie ticket.
